SSA/Projects/CIRCT/Comb/Comb.lean

```diff
@@ -0,0 +1,617 @@
+  Expr.mk
+    (op := .mul w arity)
+    (ty_eq := rfl)
+    (eff_le := by constructor)
+    (args := a)
+    (regArgs := .nil)
+
+def mux {Γ : Ctxt _} (a : Γ.Var (.bv w)) (b : Γ.Var (.bv w)) (cond : Γ.Var (.bool)) : Expr (Comb) Γ .pure (.bv w) :=
+  Expr.mk
+    (op := .mux w)
+    (ty_eq := rfl)
+    (eff_le := by constructor)
+    (args := .cons a <| .cons b <| .cons cond <| .nil)
+    (regArgs := .nil)
+
+def or {Γ : Ctxt _} (arity: Nat) (a : HVector Γ.Var (List.replicate arity (.bv w))) : Expr (Comb) Γ .pure (.bv w) :=
+  Expr.mk
+    (op := .or w arity)
+    (ty_eq := rfl)
+    (eff_le := by constructor)
+    (args := a)
+    (regArgs := .nil)
+
+def parity {Γ : Ctxt _} (a : Γ.Var (.bv w)) : Expr (Comb) Γ .pure (.bool) :=
+  Expr.mk
+    (op := .parity w)
+    (ty_eq := rfl)
+    (eff_le := by constructor)
+    (args := .cons a <| .nil)
+    (regArgs := .nil)
+
+def replicate {Γ : Ctxt _} (a : Γ.Var (.bv w)) : Expr (Comb) Γ .pure (.bv (w * n)) :=
+  Expr.mk
+    (op := .replicate w n)
+    (ty_eq := rfl)
+    (eff_le := by constructor)
+    (args := .cons a <| .nil)
+    (regArgs := .nil)
+
+def shl {Γ : Ctxt _} (a : Γ.Var (.bv w)) (b : Γ.Var (.bv w)) : Expr (Comb) Γ .pure (.bv w) :=
+  Expr.mk
+    (op := .shl w)
+    (ty_eq := rfl)
+    (eff_le := by constructor)
+    (args := .cons a <| .cons b <| .nil)
+    (regArgs := .nil)
+
+def shrs {Γ : Ctxt _} (a : Γ.Var (.bv w)) (b : Γ.Var (.bv w)) : Expr (Comb) Γ .pure (.bv w) :=
+  Expr.mk
+    (op := .shrs w)
+    (ty_eq := rfl)
+    (eff_le := by constructor)
+    (args := .cons a <| .cons b <| .nil)
+    (regArgs := .nil)
+
+def shru {Γ : Ctxt _} (a : Γ.Var (.bv w)) (b : Γ.Var (.bv w)) : Expr (Comb) Γ .pure (.bv w) :=
+  Expr.mk
+    (op := .shru w)
+    (ty_eq := rfl)
+    (eff_le := by constructor)
+    (args := .cons a <| .cons b <| .nil)
+    (regArgs := .nil)
+
+def sub {Γ : Ctxt _} (a : Γ.Var (.bv w)) (b : Γ.Var (.bv w)) : Expr (Comb) Γ .pure (.bv w) :=
+  Expr.mk
+    (op := .sub w)
+    (ty_eq := rfl)
+    (eff_le := by constructor)
+    (args := .cons a <| .cons b <| .nil)
+    (regArgs := .nil)
+
+def xor {Γ : Ctxt _} (arity: Nat) (a : HVector Γ.Var (List.replicate arity (.bv w))) : Expr (Comb) Γ .pure (.bv w) :=
+  Expr.mk
+    (op := .xor w arity)
+    (ty_eq := rfl)
+    (eff_le := by constructor)
+    (args := a)
+    (regArgs := .nil)
+
+
+/-- Convert a list of dependent pairs into an HVector -/
+def ofList {Γ : Ctxt _} ty : (l : List ((ty : Comb.Ty) × Γ.Var ty)) → (h : l.all (·.1 = ty)) → HVector (Γ.Var) (List.replicate l.length ty)
+| [], h => .nil
+| ⟨ty', var⟩::rest, h =>
+  have hty : ty' = ty := by simp_all
+  have hrest : rest.all (·.1 = ty) := by simp_all
+  .cons (hty ▸ var) (ofList _ rest hrest)
+
+def ofList' {α : Type u} {f : α → Type v} :
+    (xs : List (Σ a, f a)) → HVector f (xs.map Sigma.fst)
+  | ⟨_, x⟩ :: xs => .cons x (ofList' xs)
+  | [] => .nil
```

Comment:
This should likely go to the HVector file. Also, do we need both? They seem very similar.
If we do need both, we should have some docstring that points out the difference.

Reply:
ah no I forgot to delete one - I just wanted to keep both the implementations you and Leo proposed! will delete one.

Reply:
done